Better display of Nullables

amellnik opened the pull request:
As per #1082

amellnik commented:
As per #1082

nalimilan commented on src/abstractdataframe/io.jl:
```diff
@@ -184,8 +184,13 @@ end
         write(io, "<th>$row</th>")
         for column_name in cnames
-            cell = string(df[row, column_name])
+          if isnull(df[row, column_name])
```
You don't need this. Just call [`ourshowcompact`](https://github.com/amellnik/DataFrames.jl/blob/baeb0206fa3e0ac87fecd04d4fd48c5161e0d854/src/abstractdataframe/show.jl#L64) just like (`show`)[https://github.com/amellnik/DataFrames.jl/blob/baeb0206fa3e0ac87fecd04d4fd48c5161e0d854/src/abstractdataframe/show.jl#L255] does. Note in particular the `try...catch` block for undef entries.

Also, please use four-space indents just like in the rest of the code.

amellnik replied:
I tried `ourshowcompact` but I think I might be missing something as I'm still seeing the quotes:
![image](https://cloud.githubusercontent.com/assets/7132414/19019070/6b4b6550-882d-11e6-8477-a56741b10722.png)

nalimilan commented:
Yeah, that's how it's printed at the REPL. I think it's good to be consistent, but we can discuss whether we should print the quotes at the REPL too. Looks like it's common in data base tools, including R and Pandas data frames, to omit quotes around strings. And indeed our docs for `outshowcompact` say that quotes are supposed to be omitted. I think you can just change the method for `AbstractString` to use `print` instead of `showcompact`, and it should be good. Be sure this is tested too.

I'll need to find how to make this work well for `CategoricalArray{String}`, but that's for another PR.

amellnik commented:
Both `print` and `showcompact` include the quotation marks, so I added stripping them to a `Nullable{String}` method for `ourshowcompact`.  If this approach is OK let me know, and I'll add tests and squash.

nalimilan commented:
Please avoid special-casing as much as possible. For this, make your code as similar as possible to `show` (see the link I provided above). The different printing methods shouldn't diverge unless there's a real reason to. So please get rid of the `isnull` branch, which isn't needed since `ourshowcompact` works for null values already.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -64,6 +64,7 @@ end
 ourshowcompact(io::IO, x::Any) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::AbstractString) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::Symbol) = print(io, x) # -> Void
+ourshowcompact(io::IO, x::Nullable{String}) = isnull(x) ? print(io, "NA") : print(io, replace(get(x),  r"^\"|\"$", "")) # -> Void
```
Much simpler and cleaner: `isnull(x) ? showcompact(io, x) : print(io, get(x))`.

nalimilan commented on src/abstractdataframe/show.jl:
```diff
@@ -64,6 +64,7 @@ end
 ourshowcompact(io::IO, x::Any) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::AbstractString) = showcompact(io, x) # -> Void
```
Need to change `showcompact` to `print` if you want to remove quotes.

nalimilan commented on test/io.jl:
```diff
@@ -490,4 +490,11 @@ module TestIO
+    io = IOBuffer()
+    show(io, "text/html", df)
+    str = takebuf_string(io)
+    @test str == "<table class=\"data-frame\"><tr><th></th><th>Fish</th><th>Mass</th></tr><tr><th>1</th><td>Suzy</td><td>1.5</td></tr><tr><th>2</th><td>Amir</td><td>#NULL</td></tr></table>"
```
Can you break this into one line per row via `"""`?

amellnik replied:
On this one, the HTML output doesn't contain any linebreaks.  Is there any easy way to compare the strings ignoring `\n`?  The REPL version doesn't have this issue.

nalimilan replied:
Ah, indeed. Don't worry about it then. Just rebase and it should be good to go.

nalimilan commented on test/show.jl:
```diff
@@ -35,4 +36,11 @@ module TestShow
+    io = IOBuffer()
+    show(io, df)
+    str = takebuf_string(io)
+    @test str == "2×2 DataFrames.DataFrame\n│ Row │ Fish │ Mass  │\n├─────┼──────┼───────┤\n│ 1   │ Suzy │ 1.5   │\n│ 2   │ Amir │ #NULL │"
```
Same here.

nalimilan commented on test/io.jl:
```diff
@@ -490,4 +490,11 @@ module TestIO
     @test_throws ArgumentError eval(:(csv"foo,bar"a))
+
+    #Test HTML output for IJulia and similar
+    df = DataFrame(Fish = ["Suzy", "Amir"], Mass = [1.5, Nullable{Float64}()])
```
FWIW, you don't actually need to specify `{Float64}` since promotion does that for you.

nalimilan commented:
Thanks, but you still need to rebase against latest master.

nalimilan commented:
Test failed because of an extraneous ending `\n`.

amellnik commented:
Fixed my test.  The Latex test was failing because the Latex output was writing lines with trailing whitespace.  When I edited io.jl in Atom this whitespace was automatically trimmed, leading the test to fail.  I don't think the trailing whitespace was intentional, so I removed it and tests now pass.

nalimilan commented on src/abstractdataframe/io.jl:
```diff
@@ -232,8 +232,8 @@ function Base.show(io::IO, ::MIME"text/latex", df::AbstractDataFrame)
     write(io, "\t& ")
     header = join(map(c -> latex_escape(string(c)), cnames), " & ")
     write(io, header)
-    write(io, "\\\\ \n")
```
@maximerischard Was the space intended?

maximerischard replied:
I think it was just to make it a bit more readable than having 5 backslashes in a row.

nalimilan commented:
Thanks, sorry for the delay.

I suspect we'll have to add another special case for CategoricalArray printing. If you want to give it a look...
